src/coprocessor/dag/executor/scanner.rs

```diff
@@ -13,107 +13,125 @@
+            }
+            _ => unreachable!(),
+        };
+        Ok(())
+    }
+
+    pub fn next_row(&mut self) -> Result<Option<(Vec<u8>, Value)>> {
+        if self.no_more {
             return Ok(None);
         }
-        let scanner = self.scanner.as_mut().unwrap();
-        let kv = if self.scan_mode == ScanMode::Backward {
-            scanner.reverse_seek(Key::from_raw(&seek_key))?
-        } else {
-            scanner.seek(Key::from_raw(&seek_key))?
+
+        let kv = match self.scan_mode {
+            ScanMode::Backward => self.scanner.reverse_seek(Key::from_raw(&self.seek_key))?,
+            ScanMode::Forward => self.scanner.seek(Key::from_raw(&self.seek_key))?,
+            _ => unreachable!(),
         };
 
         let (key, value) = match kv {
             Some((key, value)) => (box_try!(key.raw()), value),
-            None => return Ok(None),
+            None => {
+                self.no_more = true;
+                return Ok(None);
+            }
         };
 
-        if range.get_start() > key.as_slice() || range.get_end() <= key.as_slice() {
+        if self.range.start > key || self.range.end <= key {
             debug!(
                 "key: {} out of range [{}, {})",
                 escape(&key),
-                escape(range.get_start()),
-                escape(range.get_end())
+                escape(self.range.get_start()),
+                escape(self.range.get_end())
             );
+            self.no_more = true;
             return Ok(None);
         }
-        Ok(Some((key, value)))
-    }
-
-    pub fn get_row(&mut self, key: &[u8]) -> Result<Option<Value>> {
-        let statistics = self.take_statistics();
-        let data = self.store.get(&Key::from_raw(key), statistics)?;
-        self.statistics = Some(statistics);
-        Ok(data)
-    }
-
-    #[inline]
-    pub fn set_seek_key(&mut self, seek_key: Option<Vec<u8>>) {
-        self.seek_key = seek_key;
-    }
 
-    pub fn init_with_range(&mut self, range: &KeyRange) -> Result<()> {
-        let upper_bound = if self.scan_mode == ScanMode::Backward {
-            self.seek_key = Some(range.get_end().to_vec());
-            None
-        } else {
-            self.seek_key = Some(range.get_start().to_vec());
-            Some(Key::from_raw(range.get_end()).encoded().to_vec())
+        self.seek_key = match (self.scan_mode, self.scan_on) {
+            (ScanMode::Forward, _) => prefix_next(&key),
+            (ScanMode::Backward, ScanOn::Table) => box_try!(truncate_as_row_key(&key)).to_vec(),
+            (ScanMode::Backward, ScanOn::Index) => key.clone(),
+            _ => unreachable!(),
         };
-        let statistics = self.take_statistics();
-        let scanner = self.store
-            .scanner(self.scan_mode, self.key_only, upper_bound, statistics)?;
-        self.scanner = Some(scanner);
-        Ok(())
+
+        CORP_GET_OR_SCAN_COUNT.with_label_values(&["range"]).inc();
```

Comment:
Now the metrics spreads everywhere, it's hard to ensure that nothing is missed or counted several times.